### DISCUSSION TITLE
Close terminal confirmation on fresh tab with no processes running

### DISCUSSION THREAD

**Discussion by RestartB** [ghostty-org/ghostty#3070] [Q&A]
Hello!

I have found in Ghostty that when I make a new terminal tab without running any commands, then close the tab, it makes me confirm that I want to do this even though no processes are running (`The terminal still has a running process. If you close the terminal the process will be killed.`). If I run a command like `echo hello world`, I am able to close the tab without confirmation as I would expect. I'm using Ghostty version `dd6460bc` with zsh and oh-my-zsh on macOS.

This screenshot shows the confirmation appearing even though nothing is running:
<img width="481" alt="Image" src="https://github.com/user-attachments/assets/287fab2a-beff-4d6d-a1c4-033bdb57c941" />

Thank you for the great work on Ghostty!

**Comment by mitchellh** [comment]
This feature relies on shell integration. This is perhaps a zsh shell integration bug. More research necessary. Converted to a discussion until we can determine what the bug is.

**Reply by RestartB** [reply]
Let me know if you need any more info and I'll get it for you 😊

**Reply by andrein** [reply]
This happens to me too. Using zsh on MacOS. Shell integration is loaded. 

If I start a new terminal, then close it, I get the same confirmation popup. 

If I run any command, or just press enter before closing, I don't get the confirmation popup.

**Reply by tbarkai** [reply]
You can confirm shell integration is loaded correctly by running:
`typeset -f _ghostty_precmd >/dev/null && echo OK || echo MISSING`

I discovered that shell integration stopped working after I installed Amazon Q, which added a bunch of lines to `.zprofile` and `.zshrc`. Once I disabled those, shell integration is working again, and `cmd+w` is closing tabs again without confirmation.

**Comment by nicolaerario** [comment]
I noticed same behavior. Simply hit `enter` can "unlock" the terminal and it can be normally closed.
I also have zsh + oh my zsh but I can see from the screenshot that `powerlevel10k` is used, like on my side. I think the problem can be (maybe) related to [instant prompt](https://github.com/romkatv/powerlevel10k?tab=readme-ov-file#instant-prompt) (?)

EDIT: I've tried disabling `Instant Prompt`; no luck 🤔

**Comment by katzkevin** [comment]
If you live on the edge,
```
confirm-close-surface = false
```

**Reply by andrwui** [reply]
or if you use tmux!!

**Reply by colinstu12** [reply]
Thanks for this! This one is in the documentation but wow the wording does not help send one there! Nothing about tabs or windows ... ("surface"??)

**Reply by stherold** [reply]
I agree this tshould rather be called one of those:
- `confirm-close-window`
- `confirm-close-ghostty`
- …

**Reply by 00-kat** [reply]
It doesn't only show up when you close the whole window nor the whole of Ghostty though. If you have two tabs, run `cat` in the first and close that tab, you'll see the prompt. If you confirm closing, it will only close that tab.

**Comment by silby** [comment]
I think that since "Confirm to close surface" relies on shell integration to detect a running process, it should default to false when shell integration isn't loaded. My shell on macos is [oksh](https://github.com/ibara/oksh/) (i.e. OpenBSD's pdksh descendant) which isn't (and probably would never be) supported by shell integration, so no matter what I do I get the confirm-close, which is a weird experience. But now I'll probably change the setting and never notice if the default changes, heh.

**Reply by 00-kat** [reply]
This should probably be mentioned in #2363.

**Comment by abtsousa** [comment]
This is happening to me as well on zsh and it seems to be a conflict with [powerlevel10k](https://github.com/romkatv/powerlevel10k), since disabling it on my [zap](https://www.zapzsh.com/) config seems to solve the issue.

**Comment by michaeleisel** [comment]
This is happening for me with bash on macOS, and happens even after I run a command or two. Is there any way to at least determine what the process is?

**Reply by michaeleisel** [reply]
Interestingly, when I use an upgraded version of bash via homebrew, and not the system one, I no longer have the issue

**Comment by guiyumin** [comment]
same here
and it seems it starts to have this annoying thing after i opened a project with cursor

**Comment by rwideman** [comment]
Not specific to zsh as this also occurs on bash but the config below works as now I dont get confirmation.
confirm-close-surface = false

**Reply by aceslick911** [reply]
Confirmed - this fixes it for me macos zsh too

**Reply by oasaleh** [reply]
Fixes it on PowerShell. Thanks!

**Comment by tahirmurata** [comment]
~Also happens on fish shell with the below config:~
```
shell-integration = fish
command = /run/current-system/sw/bin/fish
```

FIXED:
Manually setup shell integration

**Comment by Zorgatone** [comment]
Adding the following line seemed to have partially fixed it on macOS:

```
shell-integration = zsh
```

Now a newly opened tab running just zsh can be closed immediately after opening it, without confirmation.
But if I leave it open and blur, focusing on another application, then coming back to the terminal a little while later and then try to close the tab I had previously opened (without typing anything inside or running any command other than zsh on startup) I do get the confirmation popup on close

**Reply by wimpywarlord** [reply]
Can confirm this fixed it on zsh macOS with powerlevel11K

**Reply by albertobissacco** [reply]
no, it doesn't at least for me on macOs14.7 with zsh, powerlevel10K, two lines lean style +instant prompt.

**Comment by nicolaerario** [comment]
I think it can be related to something about `oh-my-zsh`: on a new MacBook (without oh-my) I don't have the problem.
Disabling oh-my-zsh on the Mac with the problem, this disappears.
I have only 2 plugins: autosuggestions and syntax highlighting (powerlevel10k as theme). I will try to install them (one by one) without oh my and see.

EDIT: it's powerlevel10k the problem (for me at least). Change it to oh-my-zsh default (robbyrussell) the terminal close every time without any confirmation dialog

**Comment by AWH1122** [comment]
I'm running a minimal zsh configuration with no theming or plugin-manager, or plugins and it is still happening to me so not sure it can be attributed directly to oh-my-zsh, or similar.

**Comment by DankeBrutus** [comment]
I am just noticing this bug(?) now as well. I thought it was some changes I made to my dot files. But it is bizarre when it shows up and when it doesn't. I get the pop-up warning about the running process in the following scenario:

- Open new tab, hit Ctrl+L or Cmd+K, then click on X to close tab or hit Cmd+W.
- This was happening on both my MacBook Pro and Mac mini.

What is really weird, to me at least, is when it does not appear. If I run even a single command, or I don't run a command at all and just hit Enter/Return, I get no warning and it closes the tab or window as I would expect it to. Additionally, if I run the `clear` command it doesn't complain if I close the tab. 

I spun up a VM of macOS Sequoia 15.6.1 to see if a fresh install would have this issue and the tabs closed just fine. Next I shared my dotfiles with the VM, made sure I installed anything missing via homebrew, then tried again. There was no issue again with closing the tabs. So I started removing or commenting out things in my dot files. I narrowed it down to these two lines:

```
source /opt/homebrew/share/zsh-syntax-highlighting/zsh-syntax-highlighting.zsh
source /opt/homebrew/share/zsh-autosuggestions/zsh-autosuggestions.zsh
```
Apparently `zsh-history-substring-search` is not part of the problem here. So it wasn't `tmux` or `fzf` like I was suspecting. For some reason syntax highlighting and autosuggestions, both separately and together,  is creating this problem with Ghostty.

edit: made the rookie mistake of writing this while still figuring stuff out. Once I had added my `.zprofile` and `.zshrc` to the VM and downloaded the ZSH packages with `brew` the same issue popped up. So that's why I combed through my `.zshrc` file.

**Comment by soulshined** [comment]
Throwing a wrench in the discussion here; I am on a fresh install of linux - fresh install of ghostty (first time trying it actually) 

My default shell is powershell - I get prompted no matter what as long as powershell is the main shell (meaning running process or any actions at all). Not sure if that's a linux or powershell thing but if you need more info from me let me know what's useful to you and I'll share

For me, I've opted to use `Ctrl-D` since it closes tabs without that prompt (in addition to panes) so I only need to know 1 shortcut for both scenarios (admittedly 'tab' in that sense means only 1 pane but that's okay for me for now). An added benefit is that `Ctrl-D` doesn't close panes with running processes so it's never an accident anyways - pretty much making the prompt's usefulness moot in my workflow. Just thought I would share that

**Reply by rhodes-b** [reply]
The proper detection requires shell-integration which there is none for powershell that is why it always prompts you
